src/OptimizedPathMappingRepository.php

```diff
@@ -0,0 +1,464 @@
+    }
+
+    /**
+     * Returns an iterator for the children paths of a resource.
+     *
+     * @param FilesystemResource $resource The resource.
+     *
+     * @return RegexFilterIterator|string[] The iterator of paths.
+     */
+    private function getChildIterator(FilesystemResource $resource)
+    {
+        $staticPrefix = rtrim($resource->getPath(), '/').'/';
+        $regExp = '~^'.preg_quote($staticPrefix, '~').'[^/]+$~';
+
+        return new RegexFilterIterator(
+            $regExp,
+            $staticPrefix,
+            new ArrayIterator($this->store->keys())
+        );
+    }
+
+    /**
+     * Returns a recursive iterator for the children paths under a given path.
+     *
+     * @param string $path The path.
+     *
+     * @return RegexFilterIterator|string[] The iterator of paths.
+     */
+    private function getRecursivePathChildIterator($path)
+    {
+        $staticPrefix = rtrim($path, '/').'/';
+        $regExp = '~^'.preg_quote($staticPrefix, '~').'.+$~';
+
+        return new RegexFilterIterator(
+            $regExp,
+            $staticPrefix,
+            new ArrayIterator($this->store->keys())
+        );
+    }
+
+    /**
+     * Returns an iterator for a glob.
+     *
+     * @param string $glob The glob.
+     *
+     * @return GlobFilterIterator|string[] The iterator of paths.
+     */
+    private function getGlobIterator($glob)
+    {
+        return new GlobFilterIterator(
+            $glob,
+            new ArrayIterator($this->store->keys())
+        );
+    }
+
+    /**
+     * Recursively creates a directory for a path.
+     *
+     * @param string $path A directory path.
+     * @return DirectoryResource The created resource
+     */
+    private function ensureDirectoryExists($path)
+    {
+        if (!$this->store->exists($path)) {
+            // Recursively initialize parent directories
+            if ($path !== '/') {
+                $this->ensureDirectoryExists(Path::getDirectory($path));
+            }
+
+            $this->store->set($path, null);
+
+            return;
+        }
+    }
+
+    /**
+     * Transform an iterator of paths into a collection of resources
+     *
+     * @param \Iterator $iterator
+     * @return FilesystemResourceCollection
+     */
+    private function iteratorToCollection(\Iterator $iterator)
+    {
+        $filesystemPaths = $this->store->getMultiple(iterator_to_array($iterator));
+        $resources = array();
+
+        foreach ($filesystemPaths as $path => $filesystemPath) {
+            $resource = $this->unserialize($filesystemPath);
+            $resource->attachTo($this, $path);
+
+            $resources[] = $resource;
+        }
+
+        return new FilesystemResourceCollection($resources);
+    }
+
+    /**
+     * Count the number of elements in the store
+     */
+    private function createRoot()
+    {
+        if ($this->store->exists('/')) {
+            return;
+        }
+
+        $this->store->set('/', null);
```

Comment:
Why do you set this to `null` and not a `GenericResource`? This way, it's impossible to do `$repo->get('/')->getChild('foo')`, no?